python/ray/profiling.py

```diff
@@ -0,0 +1,270 @@
+            # Do some computation here.
+
+    Optionally, a dictionary can be passed as the "extra_data" argument, and
+    it can have keys "name" and "cname" if you want to override the default
+    timeline display text and box color. Other values will appear at the bottom
+    of the chrome tracing GUI when you click on the box corresponding to this
+    profile span.
+
+    Args:
+        event_type: A string describing the type of the event.
+        extra_data: This must be a dictionary mapping strings to strings. This
+            data will be added to the json objects that are used to populate
+            the timeline, so if you want to set a particular color, you can
+            simply set the "cname" attribute to an appropriate color.
+            Similarly, if you set the "name" attribute, then that will set the
+            text displayed on the box in the timeline.
+
+    Returns:
+        An object that can profile a span of time via a "with" statement.
+    """
+    if worker is None:
+        worker = ray.worker.global_worker
+    if not worker.use_raylet:
+        # Log the event if this is a worker and not a driver, since the
+        # driver's event log never gets flushed.
+        if worker.mode == ray.WORKER_MODE:
+            return RayLogSpan(worker.profiler, event_type, contents=extra_data)
+        else:
+            return NULL_LOG_SPAN
+    else:
+        return RayLogSpanRaylet(
+            worker.profiler, event_type, extra_data=extra_data)
+
+
+class Profiler(object):
+    """A class that holds the profiling states.
+
+    Attributes:
+        worker: the worker to profile.
+        events: the buffer of events.
+        lock: the lock to protect access of events.
+    """
+
+    def __init__(self, worker):
+        self.worker = worker
+        self.events = []
+        self.lock = threading.Lock()
+
+    def start_flush_thread(self):
+        t = threading.Thread(target=self._periodically_flush_profile_events)
+        # Making the thread a daemon causes it to exit when the main thread
+        # exits.
+        t.daemon = True
+        t.start()
+
+    def _periodically_flush_profile_events(self):
+        """Drivers run this as a thread to flush profile data in the
+        background."""
+        # Note(rkn): This is run on a background thread in the driver. It uses
+        # the local scheduler client. This should be ok because it doesn't read
+        # from the local scheduler client and we have the GIL here. However,
+        # if either of those things changes, then we could run into issues.
+        try:
+            while True:
+                time.sleep(1)
+                self.flush_profile_data()
+        except AttributeError:
+            # This is to suppress errors that occur at shutdown.
+            pass
+
+    # TODO(rkn): Support calling this function in the middle of a task, and
+    # also call this periodically in the background from the driver.
+    def flush_profile_data(self):
```

Comment:
I think it is better to keep the flush_profile_data to the worker, and the profiling class provide a function get_profile to worker to decouple the recursive dependency.

Reply:
This function is also called by `_periodically_flush_profile_events`, if we want to move it to worker.py, we'll also need to move `_periodically_flush_profile_events` and `start_flush_thread`.

One of the main purposes of this PR is to simplify of worker.py (it's too really complex now). Thus I think it'd be better to put all profiling related code in this file. 

Does this make sense?

Reply:
The _periodically_flush_profile_events and start_flush_thread are simple,  we can keep them in the worker.  But if you think it is better to move them out of worker, i am ok to it.